### DISCUSSION TITLE
Inherit from AuthenticationStateProvider or CslaAuthenticationStateProvide

### DISCUSSION THREAD

**Discussion by ecaldentey** [MarimerLLC/csla#4016] [Questions] [answered]
Hello Everyone
I am implementing authentication and authorization with JWT and as part of the process I need to create a class that implements the abstract class AuthenticationStateProvider. However, Csla has its own implementation, which is CslaAuthenticationStateProvider. So what is the suggestion: to inherit from AuthenticationStateProvider or CslaAuthenticationStateProvider?

If the answer is CslaAuthenticationStateProvider, what are the benefits?

Thanks in advance

**Comment by rockfordlhotka** [comment]
Is this on the server or wasm? I assume wasm, because we removed any custom authn stuff from `Csla.Blazor` on the server.

On the wasm client, there is (or was) no default AuthenticationStateProvider provided generically, and so CSLA includes one with a very simple implementation - just so people can have a simple place to store the user identity.

https://github.com/MarimerLLC/csla/blob/main/Source/Csla.Blazor.WebAssembly/Authentication/CslaAuthenticationStateProvider.cs

It doesn't do anything special, and all _CSLA_ needs is that there actually _be an authn state provider_ that maintains the user principal.

CSLA uses a try pattern to add the `CslaAuthenticationStateProvider` service, so if you've added a different one already, it won't override what you've configured.

https://github.com/MarimerLLC/csla/blob/0b5048e3c2c5e0dc63c74d94ef06e6eebf7fbe28/Source/Csla.Blazor.WebAssembly/Configuration/ConfigurationExtensions.cs#L47

In summary, I'd recommend subclassing `AuthenticationStateProvider` directly.

**Reply by ecaldentey** [reply]
Hello Rocky

As I understand it, the recommended solution in Csla 8.2.2 is to have a standalone Blazor WebAssembly that implements the abstract class AuthenticationStateProvider and to have a Web API project that contains the controllers CslaStateController and DataPortalController, and implements the classes Csla.AspNetCore.Blazor.State.StateController and Csla.Server.Hosts.HttpPortalController.

Am I correct?

Thanks in advance,

Esteban

**Reply by rockfordlhotka** [reply]
There are questions of architecture here, with three general options:

1. Blazor auto rendering - so you have a Blazor server and Blazor wasm, and your pages automatically transition from server to wasm using all the latest Blazor capabilities
2. Blazor server - so everything runs on the server, possible with a mix of server-static and server-interactive pages
3. Blazor wasm - so everything runs on the wasm client

There are other variations, and for our discussion there's one more that is important:

4. Blazor wasm with a server - where the server handles server-static pages, pre-rendering of interactive pages, and all _interactive_ code runs in wasm

What I am hearing is that you are doing either number 3 or 4. Probably 3.

The CSLA 8 state management capabilities are designed to support options 1, 2, and 4. So things like the state controller and state manager are all designed for those options.

For option 3, there is no Blazor server at all, just a web server with a wasm client. By default the web server is used _only_ to download the wasm files so they can run in the browser. Your Blazor code (running in wasm) can then call services on app servers (potentially hosted on the same web server) as you choose. Normally the app server(s) will host the DataPortalController so you can use the data portal.

Because there is no server-side state or identity concept in option 3, the CslaStateController provides no value. The user can not be authenticated on the server, because you don't have any server-static pages in the app where the user could authenticate - and that's how aspnetcore would get the authenticated user principal on the server that is provided to the client using CslaStateController.

So for option 3, _you_ need to figure out how to authenticate the user securely, and how to get the results of that authentication to the wasm client where it is put into the `AuthenticationStateProvider` implementation.

**Reply by ecaldentey** [reply]
Hello Rocky,

The first thing I did was use a Blazor 8 solution with the Web App template using the Auto option, which works perfectly with CSLA. When I tried to enable JWT authentication on the client side by implementing the abstract AuthenticationStateProvider class via a CustomAuthStateProvider to handle token generation, session opening and closing, and notification of authentication state changes (NotifyAuthenticationStateChanged), I was unable to avoid what it requires, I keep receiving errors. However, I have tried unsuccessfully to implement the abstract AuthenticationStateProvider class to implement JWT authentication, as each time I cast AuthenticationStateProvider to CustomAuthStateProvider, I receive the following error, which I believe is due to how Blazor 8 hosted works. It worked perfectly in Blazor 7.

(CustomAuthStateProvider : AuthenticationStateProvider) InvalidCastException: Unable to cast object of type 'Microsoft.AspNetCore.Components.Server.ServerAuthenticationStateProvider' to type 'ICL.Catodos.ERP.CustomAuthStateProvider'.
ICL.Catodos.ERP.Server.Components.Pages.Login.Login.HandleLogin()

Because of this, I tried to replicate the Blazor 7 architecture with a Blazor 8 WebAssembly Standalone as the client and a Web AspNetCore Api project with the package Microsoft.AspNetCore.Components.WebAssembly.Server as the server to separate myself from all the rendering that Blazor 8 introduces and thus implement JWT authentication, but in this scenario, CSLA does not work for me.

Do you have any suggestions on how to resolve this situation, or is the only possible authentication option for now the one used in the ProjectTracker project?

Any help would be appreciated.

Thank you.

**Reply by rockfordlhotka** [reply]
I'm not super knowledgeable about the JWT parts of this, but I do understand the Blazor stuff quite well. Does the following make sense?

The AuthenticationStateProvider is just a way to access the current user identity from within Blazor. As far as I know, it has nothing to do with the actual authentication process, or JWT tokens, or anything like that.

Also, the only thing the authentication/identity mechanism in aspnetcore (and thus Blazor) maintains is a ClaimsPrincipal.

When your authentication service (or whatever) verifies the user's credentials, it must generate a ClaimsPrinicpal/ClaimsIdentity pair - or at least provide the data necessary for some other code to create the principal/identity pair. In your case I am guessing it also generates a JWT.

I think the simplest way to maintain the JWT is to make it a Claim in the ClaimsIdentity. That way the token should flow around the system as part of the user's principal.

Because it is a claim, you should need any special authentication state provider or anything, because only the ClaimsIdentity cares about the fact that it contains a Claim with the JWT.

**Reply by ecaldentey** [reply]
Hi Rocky,

First of all, thank you for your response. I decided to try what you suggested by saving the token in a claim and attempted to copy your authentication and authorization implementation from the ProjectTracker project, but it does not work in my Blazor 8 Web App Auto project.

In fact, I copied your exact code from the Login page, but after validating the user and adding the ClaimsPrincipal to the HttpContext, I am unable to retrieve a validated user on the Home page, even though I also copied the service configuration from your Program.cs on the server.

Could you please indicate what parameter I might be missing or what I might be doing wrong?

The only difference, I think, is that my Login page has a different layout from the default MainLayout, but I don't see how that could affect it.

This is the code I copied from ProjectTracker:

```csharp
 private async Task LoginUser()
 {
     await StateManager.InitializeAsync();
     Message = "";
     UserValidation userValidation = await portal.ExecuteAsync(userInfo.Username, userInfo.Password);
     ClaimsPrincipal principal;
     if (userValidation.IsValid)
     {
         // create authenticated principal
         var identity = new ClaimsIdentity("custom");
         var claims = new List<Claim>();
         claims.Add(new Claim(ClaimTypes.Name, userInfo.Username, ClaimValueTypes.String));
         foreach (var item in userValidation.Roles)
             claims.Add(new Claim(ClaimTypes.Role, item, ClaimValueTypes.String));
         identity.AddClaims(claims);
         principal = new ClaimsPrincipal(identity);
         var httpContext = httpContextAccessor.HttpContext;
         if (httpContext is not null)
         {
             AuthenticationProperties authProperties = new AuthenticationProperties();
             await httpContext.SignInAsync(
               CookieAuthenticationDefaults.AuthenticationScheme,
               principal,
               authProperties);

             var user = httpContextAccessor.HttpContext?.User ?? new ClaimsPrincipal(new ClaimsIdentity());
             

             NavigationManager.NavigateTo("/");
         }
     }
     else
     {
         Message = "Invalid credentials";
         //StateHasChanged();
     }
```

Thanks in advance

**Reply by rockfordlhotka** [reply]
Exactly what didn't work?

The `Login.razor` page must be server-static rendered, otherwise HttpContext isn't available (or reliable anyway).

**Reply by ecaldentey** [reply]
Hello @rockfordlhotka,

"Exactly what didn't work?"
After executing await httpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme, principal, authProperties);, 

the user never gets authenticated. My login page was not defined as SSR. Do I need to define all pages as SSR if I am going to use your implementation, or can I have pages with interactive rendering? What precautions do you suggest?

Regards

**Reply by rockfordlhotka** [reply]
I don't know why the `SignInAsync` didn't work then, that seems odd, as that should cause _all_ server-side pages (static or interactive) to be authenticated.

To get the principal to any wasm-interactive pages - that's what the CSLA state management feature can do, as demonstrated by the code in ProjectTracker.

So are you seeing that the _server-side_ pages aren't authenticated? That seems like something is wrong with the aspnetcore authn/authz configuration in your project.

Or is it _wasm client_ pages that don't have access to the user principal? That would be an issue (probably) with the configuration of the CSLA state manager feature.

**Reply by ecaldentey** [reply]
Hi Rocky,

When you created the Blazor project for ProjectTracker, what configuration did you select?
Template: Blazor Web App
Authentication type: None
Configure for Https: Yes
Interactive render mode: Auto or WebAssembly?
Interactivity location: Per page/component?

I want to replicate your exact configuration to see what might be going wrong.

Thank you very much.

**Reply by rockfordlhotka** [reply]
Blazor Web App, authz none, https yes, render mode auto, per page render mode

**Reply by ecaldentey** [reply]
Hi Rocky,

While searching for the origin of the issue why I can't save the session, I encounter two things.

If I include the instruction in my login page:
@rendermode InteractiveServer

My Csla classes stop working, meaning the following instruction cannot retrieve the Fetch method from the server side:

return await _dataPortal.FetchAsync(new FilterCriteria(userDto, DbActionTypes.Select, selectDbType, customFilters));
It doesn't find:

[Fetch]
private void Fetch(FilterCriteria filterCriteria, [Inject] ILanguageService serviceLanguage, [Inject] IUserDal dal, [Inject] IChildDataPortal<ListUserSystems> listUserSystems)
{
    UserDto userDto = new UserDto();

    Csla.Data.DataMapper.Map(filterCriteria, userDto, true);
    // ....


If I remove that instruction, Csla works perfectly again, and I can retrieve my Business Class. However, in this case, when I try to save the session information in an HttpContext or a ProtectedSessionStorage:

private async Task HandleLogin()
{
    ErrorMessage = String.Empty;

    var loginResponse = await userLoginManager.LoginUserAsync(userDto);

    if (loginResponse.LoginStatus)
    {
        await SessionStorage.SetAsync("SessionID", userDto.Username);
        // ...
    
I get the following error:

An unhandled exception occurred while processing the request.
InvalidOperationException: JavaScript interop calls cannot be issued at this time. This is because the component is being statically rendered. When prerendering is enabled, JavaScript interop calls can only be performed during the OnAfterRenderAsync lifecycle method.
Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(long asyncHandle, string identifier, string argsJson, JSCallResultType resultType, long targetInstanceId)
Am I missing any parameters to make Csla work when defining the page to render on the server side?

Regards

**Reply by rockfordlhotka** [reply]
> If I include the instruction in my login page:
> `@rendermode InteractiveServer`

The login page _must be server-static_ because it interacts with `HttpContext`, and that is only safe to do in a server-static page.

**Comment by michaelcsikos** [comment]
Hi @ecaldentey and @rockfordlhotka 

I've used the `Login.razor` as the basis to migrate from the old method of logging in with a `cshtml`. The `SignInAsync` is working, however `NavigationManager` doesn't work at all _without_ `@rendermode InteractiveServer`. It throws:
```
Exception of type 'Microsoft.AspNetCore.Components.NavigationException' was thrown.
```

I can't figure out why it works in the ProjectTracker example. If I inject a `NavigationManager` onto any SSR page, it doesn't work.

**Reply by rockfordlhotka** [reply]
I have no idea.

**Reply by michaelcsikos** [reply, marked as answer]
This is a known issue: https://github.com/dotnet/aspnetcore/issues/53996

Using an extension method `HttpContext.RedirectTo(...)`, with `[CascadingParameter] public HttpContext? HttpContext { get; set; }` and adding `method="post" Enhance` to the `EditForm` is a workaround.
```cs
public static class BlazorSsrRedirectManagerExtensions
{
    public static void RedirectTo(this HttpContext httpContext, string redirectionUrl)
    {
        ArgumentNullException.ThrowIfNull(httpContext);

        httpContext.Response.Headers.Append("blazor-enhanced-nav-redirect-location", redirectionUrl);
        httpContext.Response.StatusCode = 200;
    }
}
```
